### PR TITLE
Initial code for repository

### PR DESCRIPTION
**Pull request by mrubash1**

### CODE REVIEW

File: src/data_manipulation/datasets.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    @classmethod
+    def from_directory(cls, dirpath, thread_count, batch_size, numcep, numcontext, start_idx=0, limit=0, sort=None):
+        if not os.path.exists(dirpath):
+            raise IOError("'%s' does not exist" % dirpath)
+        txt_files = txt_filenames(dirpath, start_idx=start_idx, limit=limit, sort=sort)
+        if len(txt_files) == 0:
+            raise RuntimeError('start_idx=%d and limit=%d arguments result in zero files' % (start_idx, limit))
+        return cls(txt_files, thread_count, batch_size, numcep, numcontext)
+
+    def start_queue_threads(self, session, coord):
```
Comment (cfperez) [inline src/data_manipulation/datasets.py]: Queues are not used in this implementation and so this and 3 following related functions can be removed.

File: src/data_manipulation/datasets.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            except:
+                print("Exception occured in session.run(...)")
+                return
+
+    def next_batch(self, batch_size=None):
+        if batch_size is None:
+            batch_size = self._batch_size
+
+        (source, _, target, _,
+         self._start_idx) = my_next_batch(self._txt_files, batch_size, start_idx=self._start_idx)
```
Comment (cfperez) [inline src/data_manipulation/datasets.py]: I prefer the batching coded here and not defined in another module (i.e. `my_next_batch`). Can we pull that out?

File: src/features/utils/load_audio_to_mem.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+        past = np.reshape(past, numcontext * numcep)
+        now = orig_inputs[time_slice]
+        future = np.reshape(future, numcontext * numcep)
+
+        train_inputs[time_slice] = np.concatenate((past, now, future))
+        assert(len(train_inputs[time_slice]) == numcep + 2 * numcep * numcontext)
+
+    # Scale/standardize the inputs
+    train_inputs = (train_inputs - np.mean(train_inputs)) / np.std(train_inputs)
```
Comment (cfperez) [inline src/features/utils/load_audio_to_mem.py]: Prefer batch normalization to happen in the TensorFlow graph itself, as behavior is different between training and testing.

File: src/train_framework/tf_train_ctc.py [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+            self.loss_summary = tf.summary.scalar("avg_loss", self.avg_loss)
+
+            self.cost_placeholder = tf.placeholder(dtype=tf.float32, shape=[])
+
+            self.train_cost_op = tf.summary.scalar(
+                "train_avg_loss", self.cost_placeholder)
+
+    def setup_optimizer(self):
+        with tf.name_scope("train"):
+            self.optimizer = create_optimizer()
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: I prefer creating the optimizer here instead of behind another module: it's simpler and easier to read.
```
def setup_optimizer(self, opt_dict):
    self.optimizer = tf.train.AdamOptimizer(**opt_dict).minimize(self.avg_loss)
    return self.optimizer

File: src/train_framework/tf_train_ctc.py [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+            self.optimizer = self.optimizer.minimize(self.avg_loss)
+
+    def setup_decoder(self):
+        with tf.name_scope("decode"):
+            if self.beam_search_decoder == 'default':
+                self.decoded, self.log_prob = ctc_ops.ctc_beam_search_decoder(
+                    self.logits, self.seq_length, merge_repeated=False)
+
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: remove space

File: src/train_framework/tf_train_ctc.py [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+    def setup_summary_statistics(self):
+        # Create a placholder for the summary statistics
+        with tf.name_scope("accuracy"):
+            # Compute the edit (Levenshtein) distance of the top path
+            distance = tf.edit_distance(
+                tf.cast(self.decoded[0], tf.int32), self.targets)
+
+            # Compute the label error rate (accuracy)
+            self.ler = tf.reduce_mean(distance, name='label_error_rate')
+
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: remove space

File: src/train_framework/tf_train_ctc.py [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+        train_start = time.time()
+        for epoch in range(self.epochs):
+            # Initialize variables that can be updated
+            save_dev_model = False
+            stop_training = False
+            is_checkpoint_step, is_validation_step = \
+                self.validation_and_checkpoint_check(epoch)
+
+            # reset the index to start choosing training examples from the beginning
+            self.data_sets.train._start_idx = 0
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: How does this work with the `start_idx` config parameter? Also, prefer to not do this roundabout way of reseting the index of a "private" variable.

File: src/train_framework/tf_train_ctc.py [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+            self.train_ler_op = tf.summary.scalar(
+                "train_label_error_rate", self.ler_placeholder)
+            self.dev_ler_op = tf.summary.scalar(
+                "validation_label_error_rate", self.ler_placeholder)
+            self.test_ler_op = tf.summary.scalar(
+                "test_label_error_rate", self.ler_placeholder)
+
+    def run_training_epochs(self):
+        train_start = time.time()
+        for epoch in range(self.epochs):
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: I see `epoch+1` in various places. How about instead: `for epoch in range(1, self.epochs+1)`?

File: src/train_framework/tf_train_ctc.py [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+                self.train_ler_op, {self.ler_placeholder: self.train_ler})
+            self.writer.add_summary(summary_line, epoch)
+
+            summary_line = self.sess.run(
+                self.train_cost_op, {self.cost_placeholder: self.train_cost})
+            self.writer.add_summary(summary_line, epoch)
+
+            # Shuffle the data for the next epoch
+            if self.shuffle_data_after_epoch:
+                np.random.shuffle(self.data_sets.train._txt_files)
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: Prefer an explicit `self.data_sets_train.shuffle()` over this.

File: src/train_framework/tf_train_ctc.py [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+            if stop_training:
+                break
+
+        train_duration = time.time() - train_start
+        logger.info('Training complete, total duration: {:.2f} min'.format(
+            train_duration / 60))
+
+    def run_validation_step(self, epoch):
+        dev_ler = 0
+        print_dev_decode = True
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: This variable does nothing here--it is always True.

File: src/train_framework/tf_train_ctc.py [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+            [self.avg_loss, self.summary_op], feed)
+        self.writer.add_summary(summary_line, epoch)
+
+        return self.train_cost, self.train_ler
+
+
+# to run in console
+if __name__ == '__main__':
+
+    def main(config='neural_network.ini', name=None, debug=False):
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: Removed click, so have to add argparse for CLI. (Or put click back.)

File: src/train_framework/tf_train_ctc.py [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+        if (epoch > 0) and ((epoch + 1) != self.epochs):
+            if (epoch + 1) % self.SAVE_MODEL_EPOCH_NUM == 0:
+                is_checkpoint_step = True
+            if (epoch + 1) % self.VALIDATION_EPOCH_NUM == 0:
+                is_validation_step = True
+
+        return is_checkpoint_step, is_validation_step
+
+    def run_batches(self, dataset, is_training, decode, write_to_file, epoch):
+        n_examples = len(dataset._txt_files)
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: use `dataset.size`

File: src/data_manipulation/datasets.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+                session.run(self._enqueue_op, feed_dict={
+                    self._x: source,
+                    self._x_length: source_len,
+                    self._y: target,
+                    self._y_length: target_len})
+            except:
+                print("Exception occured in session.run(...)")
+                return
+
+    def next_batch(self, batch_size=None):
```
Comment (cfperez) [inline src/data_manipulation/datasets.py]: Implement Iterator protocol?

File: src/data_manipulation/datasets.py [hunk 14]
```diff
@@ -0,0 +1,228 @@
+import os
+from math import ceil
+from random import random
+from glob import glob
+from itertools import cycle
+from configparser import ConfigParser
+import logging
```
Comment (cfperez) [inline src/data_manipulation/datasets.py]: Add a module logger.

File: src/features/utils/load_audio_to_mem.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    '''
+
+    # Load wav files
+    fs, audio = wav.read(audio_filename)
+
+    # Get mfcc coefficients
+    orig_inputs = mfcc(audio, samplerate=fs, numcep=numcep)
+
+    # We only keep every second feature (BiRNN stride = 2)
+    orig_inputs = orig_inputs[::2]
```
Comment (cfperez) [inline src/features/utils/load_audio_to_mem.py]: Make the stride a config parameters

File: src/features/utils/load_audio_to_mem.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+    # This Source Code Form is subject to the terms of the Mozilla Public
+    # License, v. 2.0. If a copy of the MPL was not distributed with this
+    # file, You can obtain one at http://mozilla.org/MPL/2.0/.
+    '''
+
+    # Load wav files
+    fs, audio = wav.read(audio_filename)
+
+    # Get mfcc coefficients
+    orig_inputs = mfcc(audio, samplerate=fs, numcep=numcep)
```
Comment (cfperez) [inline src/features/utils/load_audio_to_mem.py]: Make `mfcc` arguments (time window and step) config parameters.

File: src/features/utils/load_audio_to_mem.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+    # Prepare pre-fix post fix context
+    empty_mfcc = np.array([])
+    empty_mfcc.resize((numcep))
+
+    # Prepare train_inputs with past and future contexts
+    time_slices = range(train_inputs.shape[0])
+    context_past_min = time_slices[0] + numcontext
+    context_future_max = time_slices[-1] - numcontext
+    for time_slice in time_slices:
```
Comment (cfperez) [inline src/features/utils/load_audio_to_mem.py]: I can't get my head around this section. Is there an easier way?

File: src/train_framework/tf_train_ctc.py [hunk 18]
```diff
@@ -0,0 +1,558 @@
...
+            self.tf_device = '/cpu:0'
+        logging.info('Using this device for main computations: %s', self.tf_device)
+
+        # set the directories
+        self.set_up_directories(model_name)
+
+        # set up the model
+        self.set_up_model()
+
+    def load_configs(self):
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: Make the training configuration its own object and pass it to Tf_train_ctc().

File: src/train_framework/tf_train_ctc.py [hunk 19]
```diff
@@ -0,0 +1,558 @@
...
+        if not os.path.exists(self.SUMMARY_DIR):
+            os.makedirs(self.SUMMARY_DIR)
+
+        # set the model name and restore if not None
+        if model_name is not None:
+            self.model_path = os.path.join(self.SESSION_DIR, model_name)
+        else:
+            self.model_path = None
+
+    def set_up_model(self):
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: This isn't really setting up the model. Maybe `set_up_data_ingest`?

File: src/train_framework/tf_train_ctc.py [hunk 20]
```diff
@@ -0,0 +1,558 @@
...
+        Batches per epoch: {}'''.format(
+            self.session_name,
+            self.n_examples_train,
+            self.n_examples_dev,
+            self.n_examples_test,
+            self.epochs,
+            self.batch_size,
+            self.n_batches_per_epoch))
+
+    def run_model(self):
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: Is there any advantage to having a `build_model()` distinct from `run_model()`? I'm wondering if for testing/prototyping, we'd like to have the finished graph that we can interact with/inspect before we start training.

File: src/train_framework/tf_train_ctc.py [hunk 21]
```diff
@@ -0,0 +1,558 @@
...
+            self.writer = tf.summary.FileWriter(
+                self.SUMMARY_DIR, graph=self.sess.graph)
+
+            # Add ops to save and restore all the variables
+            self.saver = tf.train.Saver()
+
+            # For printing out section headers
+            section = '\n{0:=^40}\n'
+
+            # If there is a model_path declared, then restore the model
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: I may also want to name a _new_ model path instead of using the default, but this doesn't let me do that.

File: src/train_framework/tf_train_ctc.py [hunk 22]
```diff
@@ -0,0 +1,558 @@
...
+                init_op = tf.global_variables_initializer()
+
+                # Initializate the weights and biases
+                self.sess.run(init_op)
+
+                # MAIN LOGIC for running the training epochs
+                logger.info(section.format('Run training epoch'))
+                self.run_training_epochs()
+
+            logger.info(section.format('Decoding test data'))
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: Maybe turn into a `_log_section(section_header)` method

File: src/train_framework/tf_train_ctc.py [hunk 23]
```diff
@@ -0,0 +1,558 @@
...
+        # Check if the current epoch is a validation or checkpoint step
+        if (epoch > 0) and ((epoch + 1) != self.epochs):
+            if (epoch + 1) % self.SAVE_MODEL_EPOCH_NUM == 0:
+                is_checkpoint_step = True
+            if (epoch + 1) % self.VALIDATION_EPOCH_NUM == 0:
+                is_validation_step = True
+
+        return is_checkpoint_step, is_validation_step
+
+    def run_batches(self, dataset, is_training, decode, write_to_file, epoch):
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: `write_to_file` isn't used.

File: src/train_framework/tf_train_ctc.py [hunk 24]
```diff
@@ -0,0 +1,558 @@
...
+                    self.input_tensor: source,
+                    self.targets: sparse_labels,
+                    self.seq_length: source_lengths}
+                )
+                dense_decoded = tf.sparse_tensor_to_dense(
+                    d, default_value=-1).eval(session=self.sess)
+                dense_labels = sparse_tuple_to_texts(sparse_labels)
+
+                # only print a set number of example translations
+                counter = 0
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]:
```
for counter in range(counter_max):
    ...
```
And drop #L524.

File: src/train_framework/tf_train_ctc.py [hunk 25]
```diff
@@ -0,0 +1,558 @@
...
+                batch_cost, _ = self.sess.run(
+                    [self.avg_loss, self.optimizer], feed)
+                self.train_cost += batch_cost * dataset._batch_size
+                logger.debug('Batch cost: %.2f | Train cost: %.2f',
+                             batch_cost, self.train_cost)
+
+            self.train_ler += self.sess.run(self.ler, feed_dict=feed) * dataset._batch_size
+            logger.debug('Label error rate: %.2f', self.train_ler)
+
+            # Turn on decode only 1 batch per epoch
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: Make the decoding its own function.

File: src/train_framework/tf_train_ctc.py [hunk 26]
```diff
@@ -0,0 +1,558 @@
...
+                        logger.info('Batch {}, file {}'.format(batch, counter))
+                        logger.info('Original: {}'.format(orig))
+                        logger.info('Decoded:  {}'.format(decoded_str))
+                        counter += 1
+
+                # save out variables for testing
+                self.dense_decoded = dense_decoded
+                self.dense_labels = dense_labels
+
+        # Metrics mean
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: these lines feel too far from related code above. Maybe extracting the decoding into a function will help, and perhaps also moving them above that.

File: src/train_framework/tf_train_ctc.py [hunk 27]
```diff
@@ -0,0 +1,558 @@
...
+        if is_training:
+            self.train_cost /= n_examples
+        self.train_ler /= n_examples
+
+        # Populate summary for histograms and distributions in tensorboard
+        self.accuracy, summary_line = self.sess.run(
```
Comment (cfperez) [inline src/train_framework/tf_train_ctc.py]: I think this session run could be combined with the one from L500.

File: src/utils/set_dirs.py [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def get_relevant_directories(
+        home_dir=None,
+        data_dir=None,
+        conf_dir=None,
+        debug=False):
+
+    home_dir = get_home_dir(home_dir=home_dir)
+
```
Comment (cfperez) [inline src/utils/set_dirs.py]: remove spaces

File: src/utils/set_dirs.py [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+def get_home_dir(home_dir=None):
+    if home_dir is None:
+        home_dir = os.environ['RNN_TUTORIAL']
+    return home_dir
+
+
+def get_data_dir(data_dir=None, home_dir=None):
+    if data_dir is None:
+        data_dir = os.path.join(get_home_dir(home_dir=home_dir), 'data', 'raw')
```
Comment (cfperez) [inline src/utils/set_dirs.py]: are you sure you want this to append 'raw'?

File: src/train_framework/tf_train_ctc.py (hunk 8 above)
Comment (mrubash1) [inline src/train_framework/tf_train_ctc.py]: This will be fixed in a future PR

File: src/data_manipulation/datasets.py [hunk 30]
```diff
@@ -0,0 +1,228 @@
...
+
+from features.utils.text import text_to_char_array, normalize_txt_file, sparse_tuple_from
+from features.utils.load_audio_to_mem import next_batch as my_next_batch
+from features.utils.load_audio_to_mem import pad_sequences
+
+DataSets = namedtuple('DataSets', 'train dev test')
+
+
+def read_datasets(conf_path, sets, numcep, numcontext,
+                  thread_count=8):
```
Comment (cfperez) [inline src/data_manipulation/datasets.py]: Also remove thread_count

**Comment by mrubash1** [comment]
Changes made as per @cfperez 's requests

**Comment by cfperez** [comment]
+2
